systemctl: hide legends with --quiet, allow overriding

keszybz opened the pull request:
This makes --quiet imply --no-legend, but allows --legend to be used to
override that. --quiet controls hints and warnings and such, and --legend controls
just the legends. I think it makes sense to allow both to controlled independently,
in particular --quiet --legend makes sense when using systemctl in a script to
provide some user-visible output.

Fixes #18560.

I think we would want to provide the same interface in other tools. I'll
convert them if there's agreement on the approach.

yuwata commented:
Looks nice!

bluca reviewed (approved):
Makes sense to me as a way to handle this combination, looks nice

poettering commented on man/standard-options.xml:
```diff
@@ -35,6 +35,16 @@
 
+  <varlistentry id='legend'>
+    <term><option>--legend</option></term>
+    <term><option>--no-legend</option></term>
```
hmm, instead of adding both, maybe just replace --no-legend with --legend=yes/no. I think this is much nicer I think sinceit means: one parameter for one internal variable, means fewer options to list in the man page (as we can then hide --no-legend) and future proof (in case --legend=auto or --legency=on-tty or so one day shall be added). 

Given that this isn't the most often used switch I think doing this via the relatively verbose =yes/no syntax is OK even if longer. It also make stuff like this from shell scripts easier: `systemctl --legend=$LEGEND …` with any typical $LEGEND value...

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -439,6 +440,7 @@ static int systemctl_parse_argv(int argc, char *argv[]) {
                 { "global",              no_argument,       NULL, ARG_GLOBAL              },
                 { "wait",                no_argument,       NULL, ARG_WAIT                },
                 { "no-block",            no_argument,       NULL, ARG_NO_BLOCK            },
+                { "legend",              no_argument,       NULL, ARG_LEGEND              },
```
as mentioned, please make --legend take a bool argument

yuwata commented:
BTW, it is welcome if shell completions for systemctl are updated.

keszybz commented:
Updated.
- shell completions as adjusted, see commit message for details.
- resolvectl and systemctl now acccept `--legend[=BOOL]`.
- I added a few preparatory patches ;)
- rebased

poettering commented on src/shared/parse-argument.c:
```diff
@@ -10,6 +10,25 @@
+        if (s) {
+                r = parse_boolean(s);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to parse %s= argument.", optname);
```
why not:

```c
return log_error_errno(r, "Failed to parse %s= argument: %s", optname, s);
```

(i.e. also say what string precisely failed parsing. people might use the same option multiple times, who knows)

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -440,7 +441,8 @@ static int systemctl_parse_argv(int argc, char *argv[]) {
                 { "wait",                no_argument,       NULL, ARG_WAIT                },
                 { "no-block",            no_argument,       NULL, ARG_NO_BLOCK            },
-                { "no-legend",           no_argument,       NULL, ARG_NO_LEGEND           },
+                { "legend",              optional_argument, NULL, ARG_LEGEND              },
```
please don't use `optional_argument`, it's seriously flawed. the thing is that `systemctl --legend yes` is suddenly ambigous. It can mean `systemctl --legend=yes list-units` or `systemctl --legend=yes yes`.

We really should never use `optional_argument`. It was a mistake starting with that ever, but going forward we should not add new ones.

The only place where one can get away with `optional_argument` is in places where we have no positional arguments at all..

Also, there's no reason to support `--legend` without argument, given that the legend is the default anyway...

Hence, I'd really just do `--legend` with `required_argument` + `--no-legend` (for compat), but stay away from `--legend` with `no_argument` and from `--legend` with `optional_argument`

keszybz replied:
Updated with the suggested changes.

poettering reviewed (changes requested):
looks good, except for the optional_argument thing

yuwata commented on man/resolvectl.xml:
```diff
@@ -339,13 +339,7 @@
-        query response are shown. Otherwise, this output is suppressed.</para></listitem>
-      </varlistentry>
-
+      <xi:include href="standard-options.xml" xpointer="legend" />
```
`resolvectl` does not have `--quiet`...

keszybz replied:
I dropped that commit. It made sense when `--legend` was allowed, but the text can just stay as i tis.

yuwata reviewed (approved):
LGTM. But one minor comment.

poettering commented on src/analyze/analyze.c:
```diff
@@ -2346,29 +2347,15 @@ static int parse_argv(int argc, char *argv[]) {
-                        } else
-                                arg_man = true;
-
+                        r = parse_boolean_argument("--man", optarg, &arg_man);
```
hmm, first argument should probably be suffixed with "=", it's a switch that expects an argument after all, and the old error message *did* suffix it like that.

i.e. we usually have this rule to suffix cmdline switches and config settings that take an argument with "=", to indicate that.

keszybz replied:
You're right. I added this now in the places where the arg is required (but not where not).

poettering commented on src/analyze/analyze.c:
```diff
@@ -2346,29 +2347,15 @@ static int parse_argv(int argc, char *argv[]) {
-                        } else
-                                arg_generators = true;
-
+                        r = parse_boolean_argument("--generators", optarg, &arg_generators);
```
ditto (here and everywhere)
